File: elixir/easy/0205/0205.exs
defmodule Solution do
  @spec is_isomorphic(s :: String.t(), t :: String.t()) :: boolean
  def is_isomorphic(s, t) do
    [size_s, size_t] =
      [s, t]
      |> Enum.map(&String.to_charlist(&1))
      |> Enum.map(&MapSet.new(&1))
      |> Enum.map(&MapSet.size(&1))

    size_zip =
      [s, t]
      |> Enum.map(&String.to_charlist(&1))
      |> Enum.zip()
      |> MapSet.new()
      |> MapSet.size()

    size_s == size_t and size_t == size_zip
  end
end

ExUnit.start()

defmodule AssertionTest do
  use ExUnit.Case, async: true

  test "is_isomorphic" do
    assert Solution.is_isomorphic("egg", "add") == true
    assert Solution.is_isomorphic("foo", "bar") == false
    assert Solution.is_isomorphic("paper", "title") == true
    assert Solution.is_isomorphic("badc", "baba") == false
  end
end

File: py/easy/0205/0205.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        dict_s = {}
        dict_t = {}
        for char_s, char_t in zip(s, t):
            if char_s not in dict_s:
                dict_s[char_s] = char_t
            else:
                if dict_s[char_s] != char_t:
                    return False
            if char_t not in dict_t:
                dict_t[char_t] = char_s
            else:
                if dict_t[char_t] != char_s:
                    return False
        return True


solution = Solution()
assert solution.isIsomorphic("egg", "add") is True
assert solution.isIsomorphic("foo", "bar") is False
assert solution.isIsomorphic("paper", "title") is True
assert solution.isIsomorphic("badc", "baba") is False
